### ISSUE TITLE
[feature request] quantized operators on CUDA

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#69364]
It's useful to be able to run a quantized transformer TorchScript model on CUDA even if some quantized operators are performed by conversion to float32. E.g. `quantized::linear_dynamic` (also requested at https://discuss.pytorch.org/t/does-dynamic-quantization-support-gpu/119231)

For TorchScript context, is it possible to manually patch `torch.ops.quantized.linear_dynamic` to support CUDA? Maybe via some manual op registration?

Thanks!

cc @jerryzh168

**Comment by jerryzh168** [comment]
Is `torch.ops.quantized.linear_dynamic` the only op you need to run on CUDA? please feel free to support this in CUDA, it can be registered with https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L522 if you have time, otherwise we may get to it later.

**Comment by jerryzh168** [comment]
We also have plans to support static quantization (QuantizedCUDA kernels) in CUDA next year

**Comment by vadimkantorov** [comment]
> the only op you need to run on CUDA

well, this was the first error. haven't expected TorchScript statically to find all such places. It would be nice if PyTorch offered some static TorchScript inspection tools: e.g. extract all op calls from a serialized model and then being able to run them through some registry and check what's supported on CUDA (without trying to actually run it)

Are dynamic quantization qparams fixed at inference time? If yes, it would be nice to have some CUDA stub ops in the mean time or a recipe/doc page of adding them manually

**Comment by jerryzh168** [comment]
I'm not exactly sure how this relates to TorchScript. `quantized::linear_dynamic` is only implemented in CPU backends currently, if we have the implementation in CUDA it would run, regardless if it is run with TorchScript or not. could you help come up with all the ops that's needed for your model?

**Comment by vadimkantorov** [comment]
You had asked what other quantized ops I need.  My usecase is to run a TorchScript-serialized quantized model. To answer this, I ran the code and got the first exception. To gather all other needed ops without progressing on the first one, I would need some static TorchScript analysis.

**Comment by jerryzh168** [comment]
can you print the TorchScript model?

**Comment by vadimkantorov** [comment]
I'll try to figure out how to print the model / computation graph. Basically, it's a transformer model

**Comment by vadimkantorov** [comment]
related: https://github.com/pytorch/pytorch/issues/74393

**Comment by vadimkantorov** [comment]
Recent CUDA quantized ops https://twitter.com/Tim_Dettmers/status/1557343499225219072 by @TimDettmers open-sourced at https://github.com/facebookresearch/bitsandbytes and https://github.com/TimDettmers/bitsandbytes

@ezyang maybe these could be upstreamed

**Comment by vadimkantorov** [comment]
The paper: https://arxiv.org/abs/2208.07339

**Comment by vadimkantorov** [comment]
ZeroQuant seems to also be studying int8 inference for transformers: https://arxiv.org/abs/2206.01861 with nice speedups

**Comment by vadimkantorov** [comment]
Another int8 training paper https://arxiv.org/abs/2304.13013 and impl in Triton:
https://github.com/TimDettmers/bitsandbytes/blob/main/bitsandbytes/triton/int8_matmul_rowwise_dequantize.py

I would say that PyTorch should now include these performant quant/dequant GPU impls + operation impls (at least gemm)

**Comment by vadimkantorov** [comment]
Related comments on int8 gemms not supported in core: https://www.speechmatics.com/company/articles-and-news/fast-and-accurate-gpu-quantization-for-transformers#part-2

**Comment by albanD** [comment]
Hey!

I think things have changed a little bit here since you opened this issue.
In particular, I would say that:
- torch.library now exists for you to extend existing native ops and create brand new ones.
- triton is shipped with pytorch so you can very easily use triton custom kernels in such new function.
- Tensor Subclasses / Modes can be used as a convenient way to add extra logic during your program's execution (for example to convert a call to aten.mm into your_lib.my_quantized_mm but also to handle quantization/dequantization and conversion between dtypes).
- cpp_extensions.load_inline allows you to write python bindings for arbitrary C++ functions on the fly with a simple string. That means you can write a binding to any cuda function you want directly from python.
- (WIP) high level custom op API to make using torch.library easier (you won't have to know the PT internals to be doing the "right thing" there).

I wonder if you had a chance to use any of these tools and if they worked for you to achieve this?
I think our current stance on a lot of these is that it is not healthy for the core pytorch team to be a "blocker" on adding any of these and we expect users to be empowered to add/extend pytorch as they see fit and we can focus our attention on techniques that cannot be done with simple extension points / require native integration. Happy to hear what you think?

**Comment by vadimkantorov** [comment]
I'll comment on other things later, but last time I tried `load_inline` (years ago ^^), it had issues with caching the kernel:
- https://github.com/pytorch/pytorch/issues/68066
in addition with quite terse docs: 
- https://github.com/pytorch/pytorch/issues/10253

**Comment by vadimkantorov** [comment]
I'll let you know my experience with bitsandbytes, but I think the whole theme of ultra-low-precision inference on CUDA is getting more attention because of LLMs. So the more primitives are included in core, the more accessible experimentation with dtypes like bits4x2 or int8 will become.

So I think, maybe an extra torch package of implementing fused quantized CUDA ops in Triton (and showcasing all these new extensions points) would be very helpful, and can borrow Triton impls from bitsandbytes and other places

**Comment by albanD** [comment]
>  So the more primitives are included in core, the more accessible experimentation with dtypes like bits4x2 or int8 will become.

There is definitely a chicken and egg problem here for us: we cannot add all of these variants to core due to limited time and concerns on library bloat. But the use of these systems is sometimes linked to us adding it. 

> And the int8 cudnn bindings in core would make it easier for the users to test their stuff against.

If this is just pointing our current convolution op for int8 inputs to cudnn, we would definitely accept a PR for that.
If this requires more exotic arguments and won't fit on the current convolution op, it will be trickier.

cc @cpuhrsch on the topic of low precision dtype integration for eexperimentation and more

**Comment by vadimkantorov** [comment]
also maybe int8,int8->int16 is useful (if supported natively by cudnn)

**Comment by vadimkantorov** [comment]
@vkuzo e.g. that blog points out about some efficient COL32 layout...

**Comment by vadimkantorov** [comment]
related: https://github.com/pytorch/pytorch/issues/74393

**Comment by vkuzo** [comment]
> that special care must be paid to layouts, and sometimes fusing an op with layout conversion (transposed output or similar?) might be important for actually bringing speedups and avoiding layout conversions inside cudnn

we really owe a detailed writeup once we have more to share, but I'll address that off the cuff...currently we are thinking of the eager mode flow as mainly focusing on numerics, and we are prioritizing performance improvements in the PT2.0 stack.  I.e. we want to enable people to get correct numerics for quantized compute in eager mode, but we may not prioritize making eager mode fast (we would welcome external contributions).  PT2.0 + inductor will be the initial prioritized path to achieve competitive performance.

**Comment by vkuzo** [comment]
yep, I agree that eager mode perf is important!  It's just a priority/bandwidth question, and we'd love help with improving the eager mode stuff once we have a POC out.

> I'm just for it being inspectable somehow to see what exactly got fused where and what memory layouts are used :)

this is possible in the PT2.0 stack today, one can enable debug settings which dumps the generated triton code which in turn contains all of that information.  It might not be the easiest to read initially, but it's all there already.  We're just ensuring all of this goodness also works for quantized compute.

**Comment by vadimkantorov** [comment]
Also, if col32 is important onwards, it would be cool to be able to express it...

**Comment by vadimkantorov** [comment]
@vkuzo it would be super cool to have a quantized multiheadattention impl / transformer encoder block :)

could be a good benchmark comparing inductor/tensorrt/bitsandbytes/something else?

**Comment by vadimkantorov** [comment]
@jerryzh168 taking detour at the original question, is there anywhere explanation in docs how exactly is dynamic quantization (and linear_dynamic in particular) working to determine qscale (and sth else?). I haven't found any details in https://pytorch.org/docs/stable/generated/torch.ao.quantization.quantize_dynamic.html?highlight=quantize_dynamic#torch.ao.quantization.quantize_dynamic or in tutorials or in the old deep dive video https://www.youtube.com/watch?v=c3MT2qV5f9w

**Comment by jerryzh168** [comment]
> @jerryzh168 taking a detour to the original question, is there anywhere explanation in docs how exactly is dynamic quantization (and linear_dynamic in particular) working to determine qscale (and sth else?). I haven't found any details in https://pytorch.org/docs/stable/generated/torch.ao.quantization.quantize_dynamic.html?highlight=quantize_dynamic#torch.ao.quantization.quantize_dynamic or in tutorials or in the old deep dive video https://www.youtube.com/watch?v=c3MT2qV5f9w
> 
> Judging from the code https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L516, it seems that it:
> 
> 1. accepts float32 as input tensor and int8 as weights
> 2. uses min-max to determine quant params for the input (using fbgemm)
> 3. quantizes the input to uint8, performs uint8 x int8 -> fp32 gemm (using mkldnn) (dequantizes output to float32 fused with matmul)
> 
> If it's so - it would be super nice to document it somewhere in docs (even if it's an old method). As something like this can be in theory easily re-simulated even in CUDA and be registered as a CUDA op. I would say that it would be nice to register something like this even as a Python op / or a C++ op but doing this naively (but correctly matching the CPU semantics)

we have some docs here: https://pytorch.org/docs/stable/generated/torch.ao.nn.quantized.dynamic.Linear.html#torch.ao.nn.quantized.dynamic.Linear but

**Comment by vadimkantorov** [comment]
So in `quantized` namespace all modules do (from float) Q-op-DQ (to float)? (I also always wonder when we can elide DQ-Q ops completely or fuse them) Are `quantizable` modules operating on torch.quint8?

Having both quantized, quantizable and intrinsic is a bit confusing haha :)

**Comment by vadimkantorov** [comment]
So I think for this summary page, it would be useful to add a few columns to the table explaining:

1) in what dtype inputs are expected / supported: e.g. float32/float16/int8/int16/quint8
2) in what dtype weights are expected / stored
3) in what dtype activations are produced / stored

**Comment by vadimkantorov** [comment]
@jerryzh168 Btw is it possible to somehow register the relevant OP linear_dynamic Python implementation and so the registration from Python as well?

I think an example of adding/patching at runtime (in eager mode) of such impls for missing backends (including missing dtype support) would be super helpful. Also if one could register in this way backward and double backward impls and could use torch.compile on them - would be super helpful for experimentation with quantized ops

**Comment by supriyar** [comment]
cc @HDCharles

**Comment by supriyar** [comment]
cc @HDCharles, who has been working on adding some primitive support in PyTorch to do int8, int8 -> int32 mm and mixed dtype matmul for weight only quantization.

**Comment by HDCharles** [comment]
> @jerryzh168 Btw is it possible to somehow register the relevant OP linear_dynamic Python implementation and so the registration from Python as well?
> 
> I think an example of adding/patching at runtime (in eager mode) of such impls for missing backends (including missing dtype support) would be super helpful. Also if one could register in this way backward and double backward impls and could use torch.compile on them - would be super helpful for experimentation with quantized ops

as of now we're kind of focusing in the opposite direction, rather than torch.compiling quantized ops like nnq.linear, there is the op torch._int_mm which can be torch.compiled into a robust triton kernel that runs in cuda. With such an option in hand, you can manually implement something like dynamic_linear in python, torch.compile the whole thing and get a reasonably performant quantized cuda op. 

We're still very early in this exploration so we haven't decided how its going to plug into the rest of the quantization APIs but as for enabling experimentation, those tools are available to play around with. Its still a very nascent area though so I don't have any public code to point to.

**Comment by HDCharles** [comment]
the main context we're focused on is getting gpu quantization working in pytorch and torch.compile + triton kernels gives us a nice way of doing that. It remains to be seen how thats going to eventually plug into the core pytorch APIs, we haven't started the design work on the final UX

**Comment by vadimkantorov** [comment]
This would be fantastic :) Btw would also be nice to add more opaque dtypes in Pytorch: https://github.com/pytorch/pytorch/issues/104702

Yeah, I understood. I mean my comment about `linear_dynamic` on GPU is also about just the dispatch system of registering by the user some inexistent op (the same problem exists outside of quantization - e.g. for integer tensors or bool ops)

**Comment by vadimkantorov** [comment]
E.g. currently

```
>>> torch._int_mm(torch.zeros(2, 2, dtype = torch.int8), torch.zeros(2, 2, dtype = torch.int8))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NotImplementedError: Could not run 'aten::_int_mm' with arguments from the 'CPU' backend. 
```

It would be cool to just in Python register Python impl of  `aten::_int_mm` for some hacks

**Comment by ezyang** [comment]
You can, use the Python op registration API :)

**Comment by vadimkantorov** [comment]
a-ha! found this https://pytorch.org/docs/stable/library.html  https://colab.research.google.com/drive/1RRhSfk7So3Cn02itzLWE9K4Fam-8U011?usp=sharing :)

**Comment by ezyang** [comment]
Yup

**Comment by vadimkantorov** [comment]
> Yup

Btw, should we be able to register in the same way `aten::slow_conv2d_cpu` for float16? (to guard against `RuntimeError: “slow_conv2d_cpu" not implemented for 'Half'` and similar NotImplemented)

From what I saw, there are now more kernels for fp16 on CPU, but until then it would be nice to be able to register some sort of "fake quantization" for fp16 which would do: dequant/cast-from-fp16-to-fp32 -> cpu op -> quant/cast-from-fp32-to-fp16

This might be important for the context of ONNX export:
- ORT supports constant folding only on CPU: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/optimizer/constant_folding.cc#L203
- we want the exported model to be fp16, so might want to be able to run the graph in fp16 without inserting in graph explicit casts/autocast contexts

This might not be important anymore with dynamo_export, but in general, an example of registering eager "fake" CPU fp16 ops might be an interesting example of using the dispatch system / torch_function

**Comment by ezyang** [comment]
It's not so easy to register a kernel for only a specific dtype, unfortunately, because we don't have a good way of going to the original kernel for regular operations lol.

**Comment by vadimkantorov** [comment]
maybe somehow could improve the dispatcher by allowing to specify dtypes explicitly in the function schema? and somehow have a `torch.call_op(schema_str, *args, **kwargs)` which can be used to call the original op using non-specific dtypes

**Comment by chengzeyi** [comment]
❤️FOR anyone who is interested in this:

I have made an implementation with `CUTLASS` and registered to the PyTorch system in my own acceleration framework `stable-fast`

This operator accepts `fp16/bf16` as input of matrix A, and `qint8` as input of matrix B, and fuses bias addition and outputs a `fp16/bf16` tensor.

It is expected to be faster than pure `fp16` kernels on newer GPUs with `fp16` act tensors. But I haven't optimized it a lot, so there is definitely a lot of space to improve it.

This is the kernel implementation:

https://github.com/chengzeyi/stable-fast/blob/74cd816b6f16cc7573755facee756c815b7b2b6a/sfast/csrc/operators/cutlass/cutlass_qlinear_dynamic_kernel.cu

This is its hacky regerestion:

https://github.com/chengzeyi/stable-fast/blob/74cd816b6f16cc7573755facee756c815b7b2b6a/sfast/csrc/operators/cutlass/cutlass_qlinear.cc

After the library is loaded, a dynamic quantized model can then run on GPUs seamlessly.

```python
@pytest.mark.parametrize('dtype', [torch.float16, torch.bfloat16])
@pytest.mark.parametrize('bias', [False, True])
@pytest.mark.parametrize('in_features', [256, 512, 1024])
@pytest.mark.parametrize('out_features', [512, 1024])
@pytest.mark.parametrize('N', [10000, 20000])
def test_benchmark_linear_dynamic(dtype, bias, in_features, out_features, N):
    with torch.no_grad():
        m = LinearModule(in_features, out_features, bias=bias).cuda().to(dtype=dtype).eval()
        m_q = torch.quantization.quantize_dynamic(m, {torch.nn.Linear},
                                                  dtype=torch.qint8)

        x = torch.randn(N, in_features).cuda().to(dtype=dtype)

        torch.cuda.synchronize()
        start = time.time()
        for _ in range(100):
            out = m(x)
        torch.cuda.synchronize()
        logger.info(f'cost={time.time() - start}')

        torch.cuda.synchronize()
        start = time.time()
        for _ in range(100):
            out = m_q(x)
        torch.cuda.synchronize()
        logger.info(f'cost={time.time() - start}')
```

**Comment by vadimkantorov** [comment]
@msaroufim @drisspg  It also appears that https://github.com/pytorch-labs/float8_experimental follows the same path as this idea above - encapsulating the functionality in some sort of replace nn.Linear or some low level functions, similar to bitsandbytes.

I think, this is a nice temporary path for experimentation with such quantized ops (as not many people understand the dispatch, so module/function-wise replacement can be a more accessible ground). And hope, some of them (or at least the low-level kernels/functions) can be included in stock pytorch

**Comment by vadimkantorov** [comment]
@ezyang For more hackability, it would also be nice to allow to do such dynamic registrations in Python (and maybe even impls in Python/Triton)

**Comment by chengzeyi** [comment]
> @ezyang For more custom tensor-subclasses/representations hackability, it would also be nice to allow to do such dynamic registrations in Python (and maybe even allow impls in Python/Triton)

If you are interested. I would tell you that I have already registered many python/triton operators in `stable-fast`.

This is the C++ code:

https://github.com/chengzeyi/stable-fast/blob/2cf3b7662ebaf8e2f016952385272171ea5ada98/sfast/csrc/jit/python_operator.cpp

This is its Python wrapper:

https://github.com/chengzeyi/stable-fast/blob/2cf3b7662ebaf8e2f016952385272171ea5ada98/sfast/utils/custom_python_operator.py

This is its usage:

https://github.com/chengzeyi/stable-fast/blob/2cf3b7662ebaf8e2f016952385272171ea5ada98/sfast/triton/torch_ops.py

If you are interested, we could talk more on this topic.
I am currently implementing SOTA optimization with diffusers and LLMs.

**Comment by ezyang** [comment]
Yeah, the original feature request still applies. It is possible to stash the old kernel function somewhere else so you can poke it, but this is definitely going to need some OSS juice. Obviously this restriction doesn't apply if you just have your own op, which is why this is not terribly high priority

**Comment by dacorvo** [comment]
This is still a work-in-progress, but you can take a look at: https://github.com/huggingface/quanto

**Comment by supriyar** [comment]
> @HDCharles @jerryzh168 @vadimkantorov im trying my best to follow the discussion and Im quite new to torch quantization modules, but it seems what is being discussed here, is that currently, there is no CUDA equivalent for eager mode dynamic quantized linear modules like https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp
> 
> 
> So then what is the path forward to have dynamic quantization for GPUs in the future? willl the user need to reimplement `dynamic_linear` and then invoke `torch compile`? Will there be plans for torch to provide a low entry barrier to a newbie who knows nothing about `torch.compile`, kernels, etc, to just use dynamic quantization, as we can now with the package `bitsandbytes`?
> 
> From this https://pytorch.org/docs/stable/quantization.html#backend-hardware-support, it seems like for GPU, dynamic quantization will not be supported.

@fabianlim you can check out the dynamic quantization offerings from torchao - https://github.com/pytorch-labs/ao?tab=readme-ov-file#a8w8-dynamic-quantization

**Comment by vadimkantorov** [comment]
I think recently, PyTorch starts to support using torch.compile'd kernels as ops. So maybe dynamic quantization on CUDA can now be stubbed/supported via this mechanism? It wouldn't be super-fast, but at least it would enable execution.

**Comment by ezyang** [comment]
@jerryzh168 do you feel that maintaining the old APIs (perhaps by pointing them to the AO library) is a legitimate use case, or should we just be telling people to switch their code to use ao library? Do we have a migration guide?

**Comment by dacorvo** [comment]
@ezyang can you elaborate what the old/new API means exactly (it is a long-standing thread).

**Comment by jerryzh168** [comment]
> @jerryzh168 do you feel that maintaining the old APIs (perhaps by pointing them to the AO library) is a legitimate use case, or should we just be telling people to switch their code to use ao library? Do we have a migration guide?

The old API is using special packed weight object while torchao is going to support this more natively in eager mode so can't really reuse what is in AO I think. For new use cases, I feel we should just tell people migrate to the new flow, dynamically quantized linear is straightforward to support in eager mode directly.

migration: I don't see a motivation to migrate old use cases, maybe we could set a deprecation deadline and deprecate the old API after our tensor subclass based flow is more mature

**Comment by jerryzh168** [comment]
> @ezyang can you elaborate what the old/new API means exactly (it is a long-standing thread).

old API is this: https://pytorch.org/tutorials/recipes/recipes/dynamic_quantization.html

new API: we are still building it in torchao, but here is a preview: https://github.com/pytorch/ao/tree/main/torchao/quantization#affine-quantization

**Comment by vadimkantorov** [comment]
One piece of feedback is that success of bitsandbytes/quanto etc shows IMO that module swap is a very comfortable/popular abstraction - at least for accuracy checks, so even if tensor subclasses are somehow mature, it would be nice to also (based on tensor subclasses under the hood) support module swaps (especially for Linear or MHA classes)

**Comment by vadimkantorov** [comment]
I wonder if it's possible to somehow polyfill `quantized::linear_dynamic` for CUDA (and maybe other legacy ops) in core PyTorch using some simple torch.compile'd impls - of course, they would not be SOTA-fast, but it would be a good baseline for the newer server-side quantization framework.

**Comment by jerryzh168** [comment]
> One piece of feedback is that success of bitsandbytes/quanto etc shows IMO that module swap is a very comfortable/popular abstraction - at least for accuracy checks, so even if tensor subclasses are somehow mature, it would be nice to also (based on tensor subclasses under the hood) support module swaps (especially for Linear or MHA classes)

makes sense, so there are two uses of tensor subclass in our design, (1) tensor subclass as data representation, basically it's an extension of the tensor dtype in pytorch core, we'd recommend this for all use cases (2) using tensor subclass as a way of code transformation, i.e. adding the dtype conversion to the code, for this one we have different ways to do it and module swap is one way people can use, even it comes with some trade offs like it will be harder to compose with other techniques like sparsity, distributed etc.

**Comment by vadimkantorov** [comment]
Yeah, I think module swaps are great as extremely easy-to-start-working-with research test bed, so it's nice to keep them supported, e.g. maybe have in core or in some library like torchao examples/scaffolding Linear8bit or MultiHeadAttention8bit or similar which would do quantized input (e.g. from fp16), do computation in smaller precision and then dequant back to fp16, and have some things like high-level util functions like quantize({nn.Linear : nn.Linear8bit}). Then people can experiment with their own quantization formats and still make use of high-level utils (and popular similar high-level modules can get upstreamed from bnb/quanto/ao to core). Of course if they show good speedups, probably a user would spend more time to quantized more of then model, fuse things up, even write custom torch.compilable functions to customize on how many bits to store per channel group / reuse lower-level primitives etc, but the simplest possible things like Linear/MHA module swaps should be supported as it's the easiest way to experiment and do research with quantization (e.g. see fp6 work)
